scipy/stats/_multivariate.py

```diff
@@ -5067,3 +5067,343 @@ def rvs(self, size=1, random_state=None):
+    distribution is completely determined by these two vectors.
+
+    Methods
+    -------
+    logpmf(x)
+        Log-probability of table `x` to occur in the distribution.
+    pmf(x)
+        Probability of table `x` to occur in the distribution.
+    mean(row, col)
+        Mean table.
+    rvs(row, col, size=1, method=None, random_state=None)
+        Draw random tables with given row and column vector sums.
+
+    Parameters
+    ----------%(_doc_callparams)s
+    %(_doc_random_state)s
+
+    Notes
+    -----
+    %(_doc_callparams_note)s
+
+    Random elements from the distribution are generated either with Boyett's
+    [1]_ or Patefield's algorithm [2]_. Boyett's algorithm has
+    O(N) time and space complexity, where N is the total sum of entries in the
+    table. Patefield's algorithm has O(K x log(N)) time complexity, where K is
+    the number of cells in the table and requires only a small constant work
+    space. By default, the `rvs` method selects the fastest algorithm based on
+    the input, but you can specify the algorithm with the keyword `method`.
+    Allowed values are "boyett" and "patefield".
+
+    .. versionadded:: 1.9.3
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.stats import random_table
+
+    >>> row = [1, 5]
+    >>> col = [2, 3, 1]
+    >>> random_table.mean(row, col)
+    array([[0.33333333, 0.5       , 0.16666667],
+           [1.66666667, 2.5       , 0.83333333]])
+
+    Alternatively, the object may be called (as a function) to fix the row
+    and column vector sums, returning a "frozen" distribution.
+
+    >>> dist = random_table(row, col)
+    >>> dist.rvs(size=1, random_state=123)
+    array([[[1., 0., 0.],
+            [1., 3., 1.]]])
+
+    References
+    ----------
+    .. [1] J. Boyett, AS 144 Appl. Statist. 28 (1979) 329-332
+    .. [2] W.M. Patefield, AS 159 Appl. Statist. 30 (1981) 91-97
+    """
+
+    def __init__(self, seed=None):
+        super().__init__(seed)
+
+    def __call__(self, row, col, *, seed=None):
+        """Create a frozen distribution of tables with given marginals.
+
+        See `random_table_frozen` for more information.
+        """
+        return random_table_frozen(row, col, seed=seed)
+
+    def logpmf(self, x):
+        """Log-probability of table to occur in the distribution.
+
+        %(_doc_pmf_callparams)s
+
+        Returns
+        -------
+        logpmf : ndarray or scalar
+            Log of the probability mass function evaluated at `x`.
+
+        %(_doc_pmf_note)s
+        """
+        x = np.asarray(x)
+
+        if np.any(x < 0):
+            raise ValueError("`x` must contain only non-negative integers")
+
+        if x.ndim < 2:
+            raise ValueError("`x` must be at least two-dimensional")
+
+        expand = x.ndim == 2
+        if expand:
+            x = x.reshape(1, *x.shape)
+
+        if x.size == 0:
+            raise ValueError("`x` must have at least one row and one column")
```

Comment:
This check seems very expensive. Can we avoid such a check, if the functions do approximately the right thing if a user passes non-integer values? The pmf is strictly only correct for integral values in the table, but the implementation accepts non-integer values because we replace the factorial with the gamma function anyway.

Reply:
Well, you could also check
```python3
%timeit not np.issubdtype(x.dtype, np.integer) or np.any(x.astype(int) != x)
```
so that it is very fast if `x` is actually of integer dtype.

Note that even if `x` is of float dtype but all integral, the wall clock time is not so different from `np.any(x < 0)` for small arrays (on my machine), and it seems to be a small fraction of the time for the actual calculation for both small and large arrays. I know you mentioned that your tables have large entries/marginals, but how big are the tables themselves (i.e. `size`)?

It is necessary by `stats` conventions, though. We'll hear about it otherwise (e.g. two weeks ago https://github.com/scipy/scipy/issues/17146#issuecomment-1271764036).

Reply:
> `np.issubdtype(x.dtype, np.integer) or np.any(x.astype(int) != x)`

Yes, that's a nice optimisation.

> wall clock time is not so different from `np.any(x < 0)`

True, that check is also expensive.

> How big are the tables themselves (i.e. size)?

It varies a lot, but we deal with 100 x 100 tables (actually histograms) and 1e6 or more entries.